whois-update/src/main/java/net/ripe/db/whois/update/mail/MailGatewaySmtp.java

```diff
@@ -82,38 +110,68 @@ public void sendEmail(final String to, final String subject, final String text,
-
-                setHeaders(mimeMessage, mimeMessage.getMessageID());
-
-                loggerContext.log("msg-out.txt", new MailMessageLogCallback(mimeMessage));
-            });
-        } catch (MailSendException e) {
+            final MimeMessage mimeMessage = mailSender.createMimeMessage();
+            setHeaders(mimeMessage);
+
+            final MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, MimeMessageHelper.MULTIPART_MODE_NO, "UTF-8");
+            helper.setFrom(mailConfiguration.getFrom());
+
+            final String punyCodedTo = PunycodeConversion.toAscii(to);
+            helper.setTo(punyCodedTo);
+
+            if (!Strings.isNullOrEmpty(replyTo)){
+                helper.setReplyTo(PunycodeConversion.toAscii(replyTo));
+            }
+
+            helper.setSubject(subject);
+            helper.setText(text);
+
+            loggerContext.log("msg-out.txt", new MailMessageLogCallback(mimeMessage));
+            mailSender.send(mimeMessage);
+            storeAsOutGoingMessage(mimeMessage, punyCodedTo);
+
+        } catch (MailSendException | MessagingException e) {
             loggerContext.log(new Message(Messages.Type.ERROR, "Caught %s: %s", e.getClass().getName(), e.getMessage()));
             LOGGER.error(String.format("Unable to send mail message to: %s", to), e);
-            if (retrySending) {
-                throw e;
+            //TODO acknowledgment should be sent even if the user is unsubscribe
+            if (mailConfiguration.retrySending() && !undeliverableMailDao.isUndeliverable(extractEmailBetweenAngleBrackets(to))) {
+                throw new MailSendException("Caught " + e.getClass().getName(), e);
             } else {
                 loggerContext.log(new Message(Messages.Type.ERROR, "Not retrying sending mail to %s with subject %s", to, subject));
             }
         }
     }
 
-    private void setHeaders(final MimeMessage mimeMessage, final String messageId) throws MessagingException {
+    private void storeAsOutGoingMessage(MimeMessage mimeMessage, String punyCodedTo) throws MessagingException {
+        outgoingMessageDao.saveOutGoingMessageId(
+            extractEmailBetweenAngleBrackets(mimeMessage.getMessageID()),   // Message-ID is in rfc2822 address format
+            extractEmailBetweenAngleBrackets(punyCodedTo));
+    }
+
+    private void setHeaders(MimeMessage mimeMessage) throws MessagingException {
         mimeMessage.addHeader("Precedence", "bulk");
         mimeMessage.addHeader("Auto-Submitted", "auto-generated");
-        mimeMessage.addHeader("List-Unsubscribe", "https://apps.db.ripe.net/db-web-ui/unsubscribe/" + messageId);
-        mimeMessage.addHeader("List-Unsubscribe-Post", "List-Unsubscribe=One-Click");
+        if (!Strings.isNullOrEmpty(mailConfiguration.getSmtpFrom())) {
+            mimeMessage.addHeader("List-Unsubscribe",
+                String.format("<%s/unsubscribe/%s>, <mailto:%s?subject=Unsubscribe%%20%s>",
```

Comment:
we can delete the slash before unsubscribe
The property is https://x/db-web-ui/